File: category-pager/src/main/java/cn/kanejin/view/CategoryPager.java
package cn.kanejin.view;

import android.content.Context;
import android.support.v4.view.ViewPager;
import android.util.AttributeSet;
import android.widget.FrameLayout;
import android.widget.RadioGroup;

/**
 * Created by Kane on 20/07/2017.
 */

public class CategoryPager extends FrameLayout {

    public CategoryPager(Context context) {
        super(context);
    }

    public CategoryPager(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public CategoryPager(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public CategoryPagerNav getNavigator() {
        return (CategoryPagerNav) findViewById(R.id.category_pager_nav);
    }

    public CategoryPagerContent getPager() {
        return (CategoryPagerContent) findViewById(R.id.category_pager_content);
    }


    public void setAdapter(CategoryPagerAdapter adapter) {
        buildNav(adapter);
        buildPager(adapter);
    }

    private void buildNav(CategoryPagerAdapter adapter) {
        CategoryPagerNav nav = getNavigator();

        if (nav == null) {
            return;
        }

        nav.setAdapter(adapter);

        nav.setOnCheckedChangeListener(new RadioGroup.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(RadioGroup group, int checkedId) {
                CategoryPagerContent pager = getPager();

                if (pager != null)
                    pager.setCurrentItem(checkedId);
            }
        });
    }

    private void buildPager(CategoryPagerAdapter adapter) {
        CategoryPagerContent pager = getPager();

        if (pager == null)
            return;

        pager.setAdapter(adapter);

        pager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {

            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(int position) {
                CategoryPagerNav nav = getNavigator();

                if (nav != null) {
                    nav.setCurrentPosition(position);
                }
            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }

        });
    }
}

File: category-pager/src/main/java/cn/kanejin/view/CategoryPager.kt
package cn.kanejin.view

import android.content.Context
import android.database.DataSetObserver
import android.support.v4.view.ViewPager
import android.util.AttributeSet
import android.widget.FrameLayout
import android.widget.RadioGroup.OnCheckedChangeListener

/**
 * Created by Kane on 20/07/2017.
 */

class CategoryPager : FrameLayout {
    val navigator: CategoryPagerNav?
        get() {
            return findViewById(R.id.category_pager_nav) as CategoryPagerNav?
        }

    val pager: CategoryPagerContent?
        get() {
            return findViewById(R.id.category_pager_content) as CategoryPagerContent?
        }

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {}

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

    fun setAdapter(adapter: CategoryPagerAdapter) {

        buildNav(adapter)

        buildPager(adapter)
    }

    private fun buildPager(adapter: CategoryPagerAdapter) {

        pager?.adapter = adapter

        adapter.registerDataSetObserver(object : DataSetObserver() {
            override fun onChanged() {
                pager?.adapter = null
                pager?.adapter = adapter
            }
        })

        pager?.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {

            override fun onPageSelected(position: Int) {
                navigator?.setCurrentPosition(position)
            }

            override fun onPageScrolled(arg0: Int, arg1: Float, arg2: Int) {

            }

            override fun onPageScrollStateChanged(arg0: Int) {
            }
        })
    }

    private fun buildNav(adapter: CategoryPagerAdapter) {
        navigator?.adapter = adapter

        navigator?.onCheckedChangeListener = OnCheckedChangeListener {
            group, checkedId ->
            pager?.currentItem = checkedId
        }
    }

}
